### PR TITLE
Package lambdas as part of `make deploy` (#1997)

### PR DESCRIPTION
**Pull request by achave11**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Review by jessebrennan — CHANGES_REQUESTED** [review]
It's not obvious to me why these are separate commits. How come you decided to split them up?

We should probably delete the `package` target in the top level makefile. I can't think of any reason we will need to run it manually. In that case we should remove all references to it. I saw at least one in the README (but there may be more).

**Comment by achave11** [comment]
> It's not obvious to me why these are separate commits. How come you decided to split them up?

I though it would be beneficial to separate additions from deletions in separate commits, but that might have been superficial and unnecessary.

File: UPGRADING.rst [hunk 1]
```diff
@@ -11,6 +11,17 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
...
+======================================================
+
+Run … ::
+
+    make deploy
+
+… to provision lambda packages and update Terraform cloud infrastructure in a
+single step.
+
+
```
Comment (jessebrennan) [inline UPGRADING.rst:24]: Consider adding a note that `make package` is defunct.

**Review by jessebrennan — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -506,33 +506,27 @@ deployment you intend to create, and once you manually provisioned
...
-to prepare the lambda functions defined in the `lambdas` directory for
-deployment via Terraform. Then run
-
 ```
 make deploy
 ```
 
-which will display a plan and ask you to confirm it. Please consult the
-Terraform documentation for details.
+to prepare the lambda functions defined in the `lambdas` directory for
```
Comment (hannes-ucsc) [inline README.md:513]:
```suggestion
to prepare the Lambda functions defined in the `lambdas` directory for
```

File: README.md [hunk 3]
```diff
@@ -506,33 +506,27 @@ deployment you intend to create, and once you manually provisioned
...
-run `make package` followed by `make deploy`.
+run `make deploy`.
 
-Some Terraform configuration is generated by `make package`, but the rest is
+Some Terraform configuration is generated by `make -C lambdas`, but the rest is
 defined in `….tf.json` files which in turn are generated from
 `….tf.json.template.py` files which are simple Python scripts containing the
 desired JSON as Python dictionary and list literals and comprehensions. Changes
 to these files require running `make deploy` again in order to update
 cloud infrastructure for the selected deployment.
```
Comment (hannes-ucsc) [inline README.md:525]:
```suggestion
Some Terraform configuration is generated by `make -C lambdas`, but the rest is
defined in `….tf.json` files which in turn are generated from
`….tf.json.template.py` templates which are simple Python scripts containing the
desired JSON as Python dictionary and list literals and comprehensions. `make deploy` will take care of both, `make -C lambda` as well as expanding the template files. Changes
to either the templates or anything in the `lambdas` directory requires running `make deploy` again in order to update
cloud infrastructure for the selected deployment.
```

File: README.md [hunk 4]
```diff
@@ -506,33 +506,27 @@ deployment you intend to create, and once you manually provisioned
...
 
-If you are provisioning multiple deployments, you will need to run `make package`
-before `make deploy` for each one, since the Terraform configuration
-generated by `make package` is deployment specific.
+If you are provisioning multiple deployments, you will need to run `make deploy`
+for each one, since the Terraform configuration generated by `make -C lambdas`
+is deployment specific.
```
Comment (hannes-ucsc) [inline README.md:529]:
```suggestion
```

That should be obvious now.

File: UPGRADING.rst [hunk 5]
```diff
@@ -11,6 +11,17 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
 
+#2802 Provision deployment by running make deploy only
+======================================================
+
+Run … ::
```
Comment (hannes-ucsc) [inline UPGRADING.rst:17]: This sounds like I have to do this to upgrade. I don't think that's the case. If this purely informational, it should be worded that way, or this upgrading entry should be removed.

File: terraform/Makefile [hunk 6]
```diff
@@ -8,7 +8,7 @@ providers: check_terraform check_branch check_aws providers.tf.json backend.tf.j
 	terraform init
 
 .PHONY: config
-config: $(patsubst %.template.py,%,$(wildcard *.tf.json.template.py))
+config: $(patsubst %.template.py,%,$(wildcard *.tf.json.template.py)) package
```
Comment (hannes-ucsc) [inline terraform/Makefile:11]:
```suggestion
config: $(patsubst %.template.py,%,$(wildcard *.tf.json.template.py)) lambdas
```

and below

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by hannes-ucsc** [comment]
Also,

image

we squash before every review.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#511review-comments

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Commit and PR titles are 1) inconsistent and 2) make no sense in light of the recent review changes.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Not much better. Please consult with peer reviewer for better commit and PR titles.

**Review by jessebrennan — APPROVED** [review]
Just a suggestion to adjust the wording.

As for a commit title, maybe "Package lambdas as part of `make deploy`"?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Back quotes missing in the PR title (I added those) and commit titles.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
No demo, because the `make deploy` target is invoked by CI/CD and frequently by other developers.

**Pull request merged**